rfd/0073-discover.md

```diff
@@ -0,0 +1,337 @@
+
+## Why
+
+The proposal is aimed at providing an easy way for Teleport administrators to
+connect their cloud-hosted resources (EC2 instances, RDS databases, EKS clusters
+an so on) to a Teleport cluster.
+
+Over the past few releases Teleport has been adding automatic discovery
+capabilities allowing it to find and register AWS databases, EC2 instances
+and (WIP as of this writing) EKS clusters. Despite providing an improved UX
+compared to registering the resources manually, connecting resources to cluster
+and setting up auto-discovery remains cumbersome with multiple different
+`teleport configure` commands and does not provide good visibility into the
+discovered resources.
+
+The `teleport discover` approach aims to take advantage of the existing auto
+discovery mechanisms Teleport has and provide a unified approach for users to
+enroll their cloud resources.
+
+## Scope
+
+- Works with both self-hosted edition and Teleport Cloud.
+- Focuses on discovering resources in AWS, other clouds to follow.
+- Covers currently supported cloud resources: EC2 instances and RDS databases.
+- The phase 1 of the implementation will only support EC2 instances and SSH.
+
+## Prerequisites
+
+In order to use `teleport discover` the user will need:
+
+- Running Auth and Proxy, or a Cloud account.
+- An AWS EC2 instance to run `teleport discover` command from. The node must have
+  IAM identity allowing it to perform discovery operations in AWS account.
+
+## Command
+
+The goal for `teleport discover` is to eventually support multiple cloud
+providers. Given that different providers often have different concepts and
+different names for similar things, it may be hard to unify them all under a
+single `teleport discover` command without making UX confusing, having
+overlapping or prefixed flags, etc.
+
+Hence, the proposal is to have a family of subcommands for different cloud
+providers:
+
+```
+$ teleport discover aws ...
+$ teleport discover gcp ...
+```
+
+This RFD will focus on `teleport discover aws`.
+
+## Usage
+
+The default mode of operation of the `teleport discover` command is to support
+the UI wizard users will use to join their resources to the cluster.
+
+Users will go through the following flow in the wizard-style dialog in the
+Teleport Web UI:
+
+- Select which environment to look for resources in. E.g. self-hosted (which
+  will use existing automatic join scipts), AWS or other cloud (which will use
+  the discover flow), as well as regions.
+- Select types of resources to import. For AWS users will be offered all types
+  we currently support auto-discovery for: EC2, RDS/Aurora, Redshift, Elasticache,
+  MemoryDB. The dialog may optionally allow to specify tags to filter the
+  resources by.
+- Based on the selected resources, display appropriate `teleport discover aws`
+  command to run and the IAM policy required for the command to successfully
+  discover and join resources.
```

Comment:
I think it's worth mentioning that we want to take an opt-out approach to resource selection, meaning that by default, we should aim to discover and import all discoverable resources. This has two effects:

1. The command `teleport discover aws` with no additional arguments should seek to discover all supported resources
2. The UI should select all resource types by default and users can deselect resource types thereby opting out

Reply:
Agree with @xinding33 here. Let's do opt-out. We can find and discover all resources and let users to opt out of resources they don't need by using labels.